client/src/lib/paymentCalc.ts

```diff
@@ -1,58 +1,108 @@
+
+  // Get overrideAxiomQueryFee from either equation or options
+  let overrideAxiomQueryFee: bigint;
+  if (options?.overrideAxiomQueryFee !== undefined) {
+    overrideAxiomQueryFee = BigInt(options.overrideAxiomQueryFee);
+  } else {
+    // overrideAxiomQueryFee = AXIOM_QUERY_FEE + projectedCallbackCost - maxFeePerGas * (callbackGasLimit + proofVerificationGas)
+    overrideAxiomQueryFee = ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI + 
+      projectedCallbackCost - maxFeePerGas * (callbackGasLimit + proofVerificationGas);
   }
-  axiomQueryFee = BigInt(axiomQueryFee ?? 0);
-  if (axiomQueryFee === 0n) {
-    axiomQueryFee = ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI;
+
+  // Get axiomQueryFee from contract if on ethereum, otherwise use L2 equation
+  let axiomQueryFee = await getAxiomV2QueryValue(
+    publicClient,
+    axiomV2QueryAddr,
+    "axiomQueryFee",
+    [],
+    ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI
+  );
+  
+  // max(overrideAxiomQueryFee, axiomQueryFee)
+  if (overrideAxiomQueryFee > axiomQueryFee) {
+    axiomQueryFee = overrideAxiomQueryFee;
   }
 
-  // Convert callback gas limit to wei
-  const callbackGasLimit = BigInt(options?.callbackGasLimit ?? ClientConstants.DEFAULT_CALLBACK_GAS_LIMIT);
+  // Calculate payment
+  let payment = axiomQueryFee + maxFeePerGas * (proofVerificationGas + callbackGasLimit);
+  const minimumPayment = projectedCallbackCost + ClientConstants.FALLBACK_AXIOM_QUERY_FEE_WEI;
+  if (payment < minimumPayment) {
+    payment = minimumPayment;
+  }
 
-  // payment = maxFeePerGas * (proofVerificationGas + callbackGasLimit) + axiomQueryFee
-  const payment =
-    BigInt(options?.maxFeePerGas ?? ClientConstants.DEFAULT_MAX_FEE_PER_GAS_WEI) *
-      (proofVerificationGas + callbackGasLimit) + axiomQueryFee;
   return payment;
 }
 
+export async function getProjectedCallbackCost(
+  chainId: string,
+  publicClient: PublicClient,
+  maxFeePerGas: bigint,
+  callbackGasLimit: bigint,
+  proofVerificationGas: bigint,
+): Promise<bigint> {
+  if (isMainnetChain(chainId)) {
+    return 0n;
+  } else if (isOpStackChain(chainId)) {
+    // on OP Stack
+    // projectedCallbackCost = 
+    //   maxFeePerGas * (callbackGasLimit + proofVerificationGas) +    
+    //   AXIOM_PROOF_CALLDATA_BYTES * 16 * (L1BlockAttributes.baseFeeScalar * L1BlockAttributes.basefee + 
+    //   L1BlockAttributes.blobBaseFeeScalar / 16 * L1BlockAttributes.blobBaseFee) / 1e6
+    const baseFeeScalar = await getOpStackL1AttributesValue(publicClient, "baseFeeScalar", []);
+    const baseFee = await getOpStackL1AttributesValue(publicClient, "basefee", []);
+    const blobBaseFeeScalar = await getOpStackL1AttributesValue(publicClient, "blobBaseFeeScalar", []);
+    const blobBaseFee = await getOpStackL1AttributesValue(publicClient, "blobBaseFee", []);
+    return maxFeePerGas * (callbackGasLimit + proofVerificationGas) + 
```

Comment:
I didn't know about this before, but found: https://viem.sh/op-stack/actions/estimateL1Gas

Reply:
I think this will be more robust than doing the computation directly